docs/design/specs/Ecma-335-Augments.md

```diff
@@ -420,3 +420,389 @@ The algorithm is amended as follows:
 
 **Section** "III.4.18 ldvirtftn" is extended to allow throwing `AmbiguousImplementationException` if the implementation of the interface method resolves at runtime to more than one default interface method. It's also extended to specify throwing `EntryPointNotFoundException` if the default interface implementation is abstract.
+
+## Static Interface Methods
+
+Follow proposed changes to the ECMA standard pertaining to static interface methods. The quotations and page numbers refer to
+version 6 from June 2012 available at:
+
+https://www.ecma-international.org/publications-and-standards/standards/ecma-335/
+
+### I.8.4.4, Virtual Methods
+
+(Add second paragraph)
+
+Static interface methods may be marked as virtual. Valid object types implementing such interfaces shall provide implementations
+for these methods by means of Method Implementations (II.15.1.4). Polymorphic behavior of calls to these methods is facilitated
+by the constrained. call IL instruction where the constrained. prefix specifies the type to use for lookup of the static interface
+method.
+
+### II.9.7, Validity of member signatures
+
+(Edit bulleted list under **Generic type definition** at the top of page 134)
+
+* Every instance method and virtual method declaration is valid with respect to S
+* Every inherited interface declaration is valid with respect to S
+* There are no restrictions on *non-virtual* static members, instance constructors or on the type's
+own generic parameter constraints.
+
+### II.9.9, Inheritance and Overriding
+
+(Edit first paragraph by adding the word *virtual* to the parenthesized formulation *for virtual **instance** methods*)
+
+Member inheritance is defined in Partition I, in “Member Inheritance”. (Overriding and hiding are also
+defined in that partition, in “Hiding, overriding, and layout”.) This definition is extended, in an obvious
+manner, in the presence of generics. Specifically, in order to determine whether a member hides (for
+static or instance members) or overrides (for virtual instance methods) a member from a base class or interface,
+simply substitute each generic parameter with its generic argument, and compare the resulting member
+signatures. [*Example*: The following illustrates this point:
+
+### II.9.11, Constrains on Generic Parameters
+
+(Change first paragraph)
+
+A generic parameter declared on a generic class or generic method can be *constrained* by one or more
+type (for encoding, see *GenericParamConstraint* table in paragraph II.22.21) and by one or more special
+constraints (paragraph II.10.1.7). Generic parameters can be instantiated only with generic arguments that are
+*assignable-to* (paragraph I.8.7.3) (when boxed) and *implements-all-static-interface-methods-of* (**paragraph
```

Comment:
Do we want to use wording used in proposed language specification at https://github.com/dotnet/csharplang/blob/main/proposals/statics-in-interfaces.md#interface-constraints-with-static-abstract-members? Specifically: "When `I` has static abstract members this needs to be further restricted so that `T` cannot itself be an interface."

Reply:
That would be an additional restriction on top of this one. This is designed to cover the case where abstract classes can be used as generic parameters.

Reply:
>That would be an additional restriction on top of this one.

Could you elaborate please?

> This is designed to cover the case where abstract classes can be used as generic parameters.

Abstract classes are not interfaces. Right?

Reply:
So, we need to restrict the generic to fully implement the constraint. Normally, this is done by requiring that the type fully implement the interface through one of two mechanisms.
1. The type must be a reference type, in which case any actual object instance must be a non-abstract type and therefore have a full implementation of the interfaces it and its base types claim to implement.
2. The type must be a valuetype which must by definition cannot be abstract type, and thus satisfy the rule above.

With the addition of the static interfaces feature, instantiations over abstract classes bring along the possibility that the type may not fully implement the interface. Thus we need to protect against that.

Now, as you note, the C# language proposal suggests that we simply prevent any interface from being used as a generic for a static interface. I would prefer to word this restriction in a different way. My preference would be to requite that the type which is used as the generic parameter to provide a full set of implementations, and to then restrict the ability to use a MethodImpl on an interface type to work with these static methods.